### PR TITLE
Update footer save warning to use the Stimulus CloneController (w-messages)

### LINKED ISSUE #10091: 🎛️ Migrate `updateFooterSaveWarning` to a Stimulus Controller `w-messages`
> ℹ️ **Part of the [Stimulus  🎛️ RFC 78](https://github.com/wagtail/rfcs/pull/78)**

### Is your proposal related to a problem?

In `client/src/entrypoints/admin/page-editor.js` there is an ad-hoc JavaScript implementation attached to the HTML `wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html` to show a warning if there are unsaved changes in the UI.

It would be good to have this code implemented as a Stimulus controller, so that there is no longer this `window.updateFooterSaveWarning` approach to make it more flexible and more testable.

### Describe the solution you'd like

Given the HTML in `wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html`...

```html
{% load wagtailadmin_tags i18n %}
<li class="footer__container footer__container--hidden" data-controller="w-unsaved" data-w-unsaved-hidden-class="footer__container--hidden" data-action="w-unsaved:update@window->w-unsaved#update">
    <div hidden data-unsaved-type="any" class="footer__save-warning footer__emphasise-span-tags">
        <div>
            <b hidden data-w-unsaved-target="edits">{% blocktrans trimmed %}You have <span>unsaved edits</span>{% endblocktrans %}</b>
            <b hidden data-w-unsaved-target="comments">{% blocktrans trimmed %}You have <span>unsaved comments</span>{% endblocktrans %}</b>
            <b hidden data-w-unsaved-target="all">{% blocktrans trimmed %}You have <span>unsaved edits</span> and <span>comments</span>{% endblocktrans %}</b>
            <p>{% trans 'Save the page before leaving' %}</p>
        </div>
    </div>
</li>

```

The existing behaviour should be preserved 100% compared to `client/src/entrypoints/admin/page-editor.js`.

As for the `window.updateFooterSaveWarning`, this should be changed to be a document dispatched event, so that the JS snippet in `wagtail/admin/templates/wagtailadmin/pages/create.html` (and edit.html) changes to;

```js
            enableDirtyFormCheck(
                '#page-edit-form',
                {
                    confirmationMessage: '{{ confirmation_message|escapejs }}',

                    {% if has_unsaved_changes %}
                        alwaysDirty: true,
                    {% endif %}
                    commentApp: window.comments.commentApp,
                    callback: (formDirty = false, commentsDirty = false) => document.dispatchEvent('w-unsaved:update', new CustomEvent({ detail { commentsDirty, formDirty }})),
                }
            );
```

#### Controller specification

* `class` hidden - replaces the hard-coding of `footer__container--hidden` in the js - see https://stimulus.hotwired.dev/reference/css-classes
* `target` edits - label target element for edits being dirty
* `target` comments - label target element for comments being dirty
* `target` all - the label element if all items are flagged
* `method` `update` (maybe a better name) this essentially just does the update as per the current JS implementation
* Values - we could store the two dirty checks as values if we want, we may want to use a data attribute to store 'active' or 'visible' though.

### Describe alternatives you've considered

* Leave as is.

### Additional context

- This is part of a larger goal to migrate to Stimulus.
- Longer term, this will set us up to tackle `enableDirtyFormCheck` probably as part of the same controller, from there we can make it easier to tackle #8453
- Not a good first issue but good for anyone semi-confident with Jest testing, JavaScript and who has taken the time to read the Stimulus docs in full.
- This is also part of a larger goal to avoid `wagtail/admin/templates/wagtailadmin/pages/_editor_js.html` being used everywhere and minimise global usage in JS See #2936
- Might be able to take a look at fixing #7188 in this issue or at least getting closer
- See https://github.com/wagtail/wagtail/tree/main/client/src/controllers for existing controller implementations

### PR DESCRIPTION
**Pull request by lb-**
* Fixes #10091
* This work is a partial enabler for #8453 (making it easier to use this behaviour in other forms) and gets us very close to being able to pull out page-editor.js (after this PR, there will only be about 4 lines of code in that file), see #2936 for context.

## Details

* The initial commit renames MessagesController to CloneController and keeps the identifier `w-messages` but also adds `w-clone`. This is because this controller is better named by the behaviour it sets up, not where it is used (making message content show), further Stimulus plans involve using this controller for its generic behaviour.
* The second commit is a refactor of the work stated in #10327 (this PR replaces that PR), the initial starting point was there but needed a fair bit of work to get to a functional state.
- Controller changes include
  - Changes default show-delay to zero (immediate) and update the w-messages (page messages) to have the 100ms delay
  - Add support for 'hide' classes to be added/removed based on clone usage
  - Add support for calling the clear method from actions, and resetting classes if used after an animation duration or max delay
  - Pull out logic to get template content to class method
  - General clean up of the original messages only implementation to be better documented, typed and tested, bringing in the learnings from other Controller work
- Refine the `debounce` util to be more useful, updating Types to be better, now returns a Promise API as per the planned implementation for #10356 and also allows this to be used without a delay. This means the one util can be used to conditionally immediately run the callback. Super useful when we want the same code to run maybe delayed as we are doing in a few of the Stimulus controllers.
- Finally, this adopts the w-messages (CloneController) in unsaved changes warning, replacing the need for the global `window.updateFooterSaveWarning` and now uses Event dispatching to perform the same behaviour.
- I have added `aria-live="polite"` so that screen readers may advise users of these messages.

## Notes on scope

* Any core functional changes to the `enableDirtyFormCheck` function, this is what runs to compare the initial form state to the potentially changed state and sets up comment changes listeners. There are known issues with this code but at the moment this PR is not trying to refine/resolve any of them, behaviour should just be the same as current state.
* Backwards compatible global `updateFooterSaveWarning`, can add this if needed, but this felt like an internal function and would not be called by custom code.
* I added `{% block message_templates %}`, I thought this may be a nice way to add more unsaved pages messages through customisations, this is intentionally not documented.

## Checklist

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour? (N/A)
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments? Chrome 118, Firefox 115, Safari 16.1
-   [X] For new features: Has the documentation been updated accordingly? (added release considerations

## Testing

Check these items in both a new Page creation form and an existing Page edit form. Note that the page create form acts a bit differently with how it calculates whether the page has edits. See also #7188 .

1. Check that when the page form loads, after ~3 seconds, no rogue 'unsaved edits' appears except for know issue #1234
2. Now make an edit to any basic field, observe that the unsaved field edits message appears in the footer
3. Now add a comment to any field, observe the message should change to field & comments in the footer
4. Remove the comment, observe the message updates
5. You can also remove the change made in step 2 (basic field change), set it back to the previous, the message should remove after (~2s)
6. Refresh, do not save anything, now just add a comment somewhere. Observe the message should appear.
7. Now remove that comment, observe after a delay (~2s) the message should be removed
8. **additional test** Add a comment, observe the message shows, now remove the comment and quickly add one, wait ~2 seconds and confirm the message is still shown. (this checks the setTimeout tracking).

### CODE REVIEW

File: client/src/controllers/CloneController.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+   */
+  getTemplateContent(type?: string | null): HTMLElement | null {
+    const template =
+      (type &&
+        this.templateTargets.find(({ dataset }) => dataset.type === type)) ||
+      this.templateTarget;
+    const content = template.content.firstElementChild?.cloneNode(true);
+    return content instanceof HTMLElement ? content : null;
+  }
+}
```
Comment (lb-) [inline client/src/controllers/CloneController.ts:148]: Note: Unfortunately Git diffs do not do renamed files well. However the MessagesController has been renamed and while a lot has changed the core behaviour is still there. When reviewing the commits locally it should be a bit clearer what changed.

File: wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html [hunk 2]
```diff
@@ -1,11 +1,32 @@
...
+<li
+    class="footer__container footer__container--hidden"
+    role="status"
+    data-controller="w-messages"
+    data-action="w-unsaved:add@document->w-messages#add w-unsaved:clear@document->w-messages#clear"
+    data-w-messages-clear-param="true"
+    data-w-messages-hide-class="footer__container--hidden"
+>
+    <div class="footer__save-warning footer__emphasise-span-tags" data-w-messages-target="container"></div>
+    {% trans 'Save the page before leaving' as generic_save_page_message %}
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html:11]: Note: If we want, I can pull this 'shared' label up into the data-w-messages-text-param and remove from each of the `p` tags below. Not sure if that will make this HTML readability worse though.

File: wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html [hunk 3]
```diff
@@ -1,11 +1,32 @@
...
-        <div>
-            <b hidden data-unsaved-type="edits">{% blocktrans trimmed %}You have <span>unsaved edits</span>{% endblocktrans %}</b>
-            <b hidden data-unsaved-type="comments">{% blocktrans trimmed %}You have <span>unsaved comments</span>{% endblocktrans %}</b>
-            <b hidden data-unsaved-type="all">{% blocktrans trimmed %}You have <span>unsaved edits</span> and <span>comments</span>{% endblocktrans %}</b>
-            <p>{% trans 'Save the page before leaving' %}</p>
-        </div>
-    </div>
+<li
+    class="footer__container footer__container--hidden"
+    role="status"
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html:4]:
```suggestion
    aria-live="polite"
```

**Comment by lb-** [comment]
Just rebased this and removed some code that was not really required for this feature adoption.

File: wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html (hunk 2 above)
Comment (laymonage) [reply wagtail/admin/templates/wagtailadmin/pages/_unsaved_changes_warning.html:12]: I don't think it's necessary, though perhaps it should be inside the `message_templates` block below, so if we later extract this into a shared template, we'll literally just have the block's contents in this file, e.g.

```html+django
{% extends "wagtailadmin/shared/unsaved_changes_warning.html" %}
{% block message_templates %}
    {% trans 'Save the page before leaving' as generic_save_page_message %}
    {# The rest of the templates here ... #}
{% endblock %}
```

File: client/src/entrypoints/admin/page-editor.js [hunk 4]
```diff
@@ -24,43 +24,13 @@ function initKeyboardShortcuts() {
 
 window.initKeyboardShortcuts = initKeyboardShortcuts;
 
+/**
+ * Ensure any code calling this global function does not throw an error.
+ *
+ * @deprecated RemovedInWagtail60
+ */
+window.updateFooterSaveWarning = () => {};
```
Comment (laymonage) [inline client/src/entrypoints/admin/page-editor.js:32]: > Backwards compatible global `updateFooterSaveWarning`, can add this if needed, but this felt like an internal function and would not be called by custom code.

I don't think it's necessary, this stub function is more than enough 👍

**Review by laymonage — APPROVED** [review]
Tested all the listed scenarios and happy with the overall changes. Thanks for keeping the PR scope focused so it's easy to review!

Just have a minor comment on moving the translatable message string to inside the newly-introduced block. Otherwise it's all good 👍

File: client/src/entrypoints/admin/page-editor.js (hunk 4 above)
Comment (lb-) [reply client/src/entrypoints/admin/page-editor.js:32]: Great, I will remove this.

**Comment by lb-** [comment]
Thanks @laymonage  I will merge this in.

I have moved that translation into inside template block, good call.

I have removed the noop stub of the removed function and updated the release considerations wording accordingly.

Added changelog and will merge once the CI passes.

Thanks @Hussain0520 for your original start here, really appreciate it.

**Pull request merged**